tour_model_eval/load_predict.py

```diff
@@ -0,0 +1,92 @@
+import emission.core.get_database as edb
+import emission.analysis.modelling.tour_model.similarity as similarity
+import emission.analysis.modelling.tour_model.data_preprocessing as preprocess
+import jsonpickle as jpickle
+
+
+def loadModelStage(filename):
+    import emission.analysis.classification.inference.mode.seed.pipeline as seedp
+    model = loadModel(filename)
+    return model
+
+
+def loadModel(filename):
+    fd = open(filename, "r")
+    all_model = fd.read()
+    all_model = jpickle.loads(all_model)
+    fd.close()
+    return all_model
+
+
+def in_bin(bin_location_features,new_trip_location_feat,radius):
+    start_b_lon = new_trip_location_feat[0]
+    start_b_lat = new_trip_location_feat[1]
+    end_b_lon = new_trip_location_feat[2]
+    end_b_lat = new_trip_location_feat[3]
+    for feat in bin_location_features:
+        start_a_lon = feat[0]
+        start_a_lat = feat[1]
+        end_a_lon = feat[2]
+        end_a_lat = feat[3]
+        start = similarity.within_radius(start_a_lat, start_a_lon, start_b_lat, start_b_lon,radius)
+        end = similarity.within_radius(end_a_lat, end_a_lon, end_b_lat, end_b_lon, radius)
+        if start and end:
+            continue
+        else:
+            return False
+    return True
+
+
+def predict_labels(trip):
+    radius = 100
+    user = trip[0]['user_id']
+    trip_feat = preprocess.extract_features(trip)[0]
+    trip_loc_feat = trip_feat[0:4]
+    try:
+        # load locations of bins(1st round of clustering)
+        bin_locations = loadModelStage('locations_' + str(user))[0]
+        # load models from the 2nd round of clustering
+        models = loadModelStage('models_' + str(user))[0]
+        # load user labels in all clusters
+        user_labels = loadModelStage('user_labels_' + str(user))[0]
+    except IOError:
+        return {}
+
+    first_round_label_set = list(bin_locations.keys())
+    sel_fl = None
+    for fl in first_round_label_set:
+        # extract location features of selected bin
+        sel_loc_feat = bin_locations[fl]
+        # Check if start/end locations of the new trip and every start/end locations in this bin are within the range of
+        # radius. If so, the new trip falls in this bin. Then predict the second round label of the new trip
+        # using this bin's model
+        if in_bin(sel_loc_feat, trip_loc_feat, radius):
+            sel_fl = fl
+            break
```

Comment:
This is too intensive for prediction. Basically, for prediction, you do **not** want to write any code that iterates over all the trips for a user; it will not scale.

It will not scale means that if you have a user with data for 6 months, the server will crash or it will take hours for the computation.

Reply:
I didn't actually run all trips. I saved all coordinates for all common trips in the `build and save` step. My logic here is to compare the new trip's coordinates with other trips'  to see which bin it should fall in. 
The other way I can think of is that finding a representative trip for every bin, then compare it with the new trip. But there is a risk that new trip match a representative trip but not all the existing trips in that bin. ( I thought that the new trips would match all trips in a bin if it matched one, but that would not be certain.)
Do you have other way to see if a new trip can fall in any bin?

Reply:
There is no concept of running a trip. A trip is data, you can only run code.
I think that you mean that you are not running the clustering algorithm on all trips.
You are not running the clustering algorithm, but you are still running a `for` loop over all the trips aka "iterating over the trips". Iterating over 10k trips to check the distances is still going to take time.

Using a representative trip has the problem that you have outlined " I thought that the new trips would match all trips in a bin if it matched one, but that would not be certain". A better option seems to be to store a geometric shape representing all the starts and all the ends of all the trips. You can then quickly check whether the new trip is within the shape. 

The simplest shape is probably a circle. A potential algorithm might be:
- get all the points
- find the centroid
- find the max distance of any point from the centroid
- use that as the circle representing the points

You can also use a minimum enclosing circle algorithm/implementation instead of the hacky solution above.
https://www.nayuki.io/page/smallest-enclosing-circle

Reply:
Actually, you can't use that implementation because its license is LGPL and we don't want to taint our BSD-3 license.
An alternative is OpenCV, which uses an Apache license.

Reply:
A lighter weight alternative appears to be
https://github.com/luwei14/ShapeAnalyzer

Make sure to update the `ThirdPartyLicenses.md` file if you include a new library

Reply:
This has a super simple example using scipy
https://www.cs.au.dk/~gerth/ipsa18/slides/optimization.pdf